### PR TITLE
🐎 Only update ticket with tags, be optimistically concurrent

### PR DESCRIPTION
**Pull request by benagain**
Prevent a race condition with other systems updating the ticket at the same time.  See #32.

### CODE REVIEW

File: src/SFA.DAS.Zendesk.Monitor/Zendesk/Model/RootObjectConverter`1.cs [hunk 1]
```diff
@@ -0,0 +1,67 @@
+﻿using Newtonsoft.Json;
```
Comment (DannyGB) [inline src/SFA.DAS.Zendesk.Monitor/Zendesk/Model/RootObjectConverter`1.cs]: Why has the filename got a back tick in it? I'd remove it if possible

File: src/SFA.DAS.Zendesk.Monitor/Zendesk/Model/SafeModifyTags.cs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+#pragma warning disable CA2208 // Instantiate argument exceptions correctly
+            UpdatedStamp = current?.UpdatedAt
+                ?? throw new ArgumentNullException(nameof(Ticket.UpdatedAt));
+
+            Tags = current?.Tags
+                ?? throw new ArgumentNullException(nameof(Ticket.Tags));
+        }
+
+        public bool SafeUpdate { get; } = true;
```
Comment (DannyGB) [inline src/SFA.DAS.Zendesk.Monitor/Zendesk/Model/SafeModifyTags.cs]: Is the goal here to make this class thread-safe or something else? If it is there are probably better ways to do it.

File: src/SFA.DAS.Zendesk.Monitor.UnitTests/Tests/ModelSerialisationTests.Zendesk.cs [hunk 3]
```diff
@@ -1,13 +1,18 @@
...
 using System;
+using System.Collections.Generic;
 using System.Linq;
 using Xunit;
 
 namespace SFA.DAS.Zendesk.Monitor.UnitTests
 {
     public class ModelSerialisationTests
     {
+        private const string iso8601 = "yyyy-MM-ddTHH:mm:ss.FFFFFFFK";
```
[the remark this answers is not in the document]
Comment (benagain) [reply src/SFA.DAS.Zendesk.Monitor.UnitTests/Tests/ModelSerialisationTests.Zendesk.cs]: Almost.  It is ISO8601 but trailing zeros in the seconds are not included.  E.g.

    var d = DateTimeOffset.Parse("2018-03-05T05:23:50.8298130+00:00");
    Console.WriteLine(d.ToString("o"));
    Console.WriteLine(d.ToString("yyyy-MM-ddTHH:mm:ss.FFFFFFFK"));

Outputs 

    2018-03-05T05:23:50.8298130+00:00
    2018-03-05T05:23:50.829813+00:00

This is used simply to verify that the object serialised by JsonConvert is correct, and that format string is the [one used by JsonConvert](https://github.com/JamesNK/Newtonsoft.Json/blob/master/Src/Newtonsoft.Json/Utilities/DateTimeUtils.cs).

File: src/SFA.DAS.Zendesk.Monitor.UserAcceptanceTests/Fakes/FakeZendesk.cs [hunk 4]
```diff
@@ -91,10 +91,8 @@ internal async Task<Ticket> CreateTicket()
...
         }
 
-        internal Task UpdateTicket(Ticket ticket)
-        {
-            return zendeskApi.PutTicket(ticket.Id, new TicketRequest { Ticket = ticket });
-        }
+        internal Task AddTag(Ticket ticket, string v)
+            => zendeskApi.ModifyTags(ticket, additions: new[] { v });
 
         public Task<long[]> /*ISharingTickets.*/GetTicketsForSharing() => sharing.GetTicketsForSharing();
```
[the remark this answers is not in the document]
Comment (benagain) [reply src/SFA.DAS.Zendesk.Monitor.UserAcceptanceTests/Fakes/FakeZendesk.cs]: I made it an explicit implementation 👌

File: src/SFA.DAS.Zendesk.Monitor/Zendesk/Model/RootObjectConverter`1.cs (hunk 1 above)
Comment (benagain) [reply src/SFA.DAS.Zendesk.Monitor/Zendesk/Model/RootObjectConverter`1.cs]: It's the name Visual Studio gave to the file, seems to be the MS de-facto standard for naming files containing generic classes.

I've renamed it to `RootObjectConverter_1.cs` which hopefully removes the backtick shudder.

File: src/SFA.DAS.Zendesk.Monitor/Zendesk/Model/SafeModifyTags.cs (hunk 2 above)
Comment (benagain) [reply src/SFA.DAS.Zendesk.Monitor/Zendesk/Model/SafeModifyTags.cs]: Not really, this is to make concurrent updates from multiple systems safer by preventing updates to a record if that record has been modified between reading and updating; the update must be retried after a read of the new state of the record.

Essentially it's Zendesk's version of [optimistic concurrency](https://en.wikipedia.org/wiki/Optimistic_concurrency_control).

File: src/SFA.DAS.Zendesk.Monitor.UserAcceptanceTests/MiddlewareReceiveTicketUpdatesSteps.cs [hunk 5]
```diff
@@ -38,8 +38,7 @@ public async Task GivenATicketExists()
         [When(@"the ticket is marked to be shared")]
         public async Task WhenTheTicketIsMarkedToBeShared()
         {
-            data.Ticket.Tags.Add("pending_middleware");
-            await zendesk.UpdateTicket(data.Ticket);
+            await zendesk.AddTag(data.Ticket, "pending_middleware");
```
[the remark this answers is not in the document]
Comment (benagain) [reply src/SFA.DAS.Zendesk.Monitor.UserAcceptanceTests/MiddlewareReceiveTicketUpdatesSteps.cs]: 👌

**Pull request merged**